src/components/media-loader.js

```diff
@@ -53,62 +55,72 @@ AFRAME.registerComponent("media-loader", {
     const loadingObj = new THREE.Mesh(new THREE.BoxGeometry(), new THREE.MeshBasicMaterial());
     this.el.setObject3D("mesh", loadingObj);
     this.setShapeAndScale(true);
+    delete this.showLoaderTimeout;
   },
 
-  // TODO: correctly handle case where src changes
-  async update() {
+  async update(oldData) {
     try {
-      const url = this.data.src;
-      const token = this.data.token;
+      const { src, index } = this.data;
 
-      this.showLoaderTimeout = this.showLoaderTimeout || setTimeout(this.showLoader, 100);
-
-      if (!url) return;
+      if (src !== oldData.src && !this.showLoaderTimeout) {
+        this.showLoaderTimeout = setTimeout(this.showLoader, 100);
+      }
 
-      const { raw, contentType } = await resolveMedia(url, token);
+      if (!src) return;
 
-      if (token) {
-        if (this.blobURL) {
-          URL.revokeObjectURL(this.blobURL);
-          this.blobURL = null;
-        }
-        const response = await fetch(raw, {
-          method: "GET",
-          headers: { Authorization: `Token ${token}` }
-        });
-        const blob = await response.blob();
-        this.blobURL = window.URL.createObjectURL(blob);
-      }
+      const { raw, images, contentType } = await resolveMedia(src, false, index);
 
-      if (contentType.startsWith("image/") || contentType.startsWith("video/") || contentType.startsWith("audio/")) {
+      const isPDF = contentType.startsWith("application/pdf");
+      if (
+        contentType.startsWith("image/") ||
+        contentType.startsWith("video/") ||
+        contentType.startsWith("audio/") ||
+        isPDF
+      ) {
+        this.el.removeAttribute("gltf-model-plus");
         this.el.addEventListener(
           "image-loaded",
-          () => {
+          async () => {
             clearTimeout(this.showLoaderTimeout);
+            delete this.showLoaderTimeout;
+            if (isPDF) {
+              const maxIndex = await fetchMaxContentIndex(src, images.png);
```

Comment:
Is there a reason to prefer either `Promise.prototype.then` or (`async` and `await`) when either is available?

Reply:
Not really no. I usually use then if I can achieve the result in a point free way, or with a simple funciton with implicit return like `fetch().then(r => r.json())` but use async for anything else unless its particularly awkward for some reason like we saw with the constructor example Marshal shoed us.